src/libmongoc/src/mongoc/mongoc-secure-channel.c

```diff
@@ -207,14 +222,67 @@ decode_object (const char *structType,
    return out;
 }
 
+// `generate_key_name` generates a deterministic name for a key of the form: "libmongoc-<SHA256 fingerprint>-<suffix>".
+// Returns NULL on error.
+static LPWSTR
+generate_key_name (LPBYTE data, DWORD len, const char *suffix)
+{
+   bool ok = false;
+   char *hash_hex = NULL;
+   char *key_name = NULL;
+   LPWSTR key_name_wide = NULL;
+
+   BSON_ASSERT_PARAM (data);
+   BSON_ASSERT_PARAM (suffix);
+
+   // Compute a hash of the certificate:
+   {
+      unsigned char hash[32];
+      mongoc_crypto_t crypto;
+      mongoc_crypto_init (&crypto, MONGOC_CRYPTO_ALGORITHM_SHA_256);
+      BSON_ASSERT (mlib_in_range (size_t, len));
+      if (!mongoc_crypto_hash (&crypto, (const unsigned char *) data, (size_t) len, hash)) {
+         goto fail;
+      }
+      // Use uppercase hex to match form of `openssl x509` command:
+      hash_hex = bin_to_hex_uppercase ((const uint8_t *) hash, sizeof (hash));
+      if (!hash_hex) {
+         goto fail;
+      }
+   }
+
+   // Convert to a wide string:
+   {
+      key_name = bson_strdup_printf ("libmongoc-%s-%s", hash_hex, suffix);
+      size_t key_name_wide_chars = strlen (key_name) + 1;
+      key_name_wide = bson_malloc (sizeof (WCHAR) * (key_name_wide_chars));
+      BSON_ASSERT (mlib_in_range (int, key_name_wide_chars));
+      if (0 == MultiByteToWideChar (
```

Comment:
`MultiByteToWideChar` (and other Win32 string conversion APIs) is a foot bazooka, and I get nervous every time I see it. In this case, it's _probably_ safe, but will possibly allocate more `WCHAR` than necessary.

We're only using it in a few places, but it may be good to refactor it into an internal API that "does the right thing" every time rather than needing to do the math and audit it every time.